pyccel/parser/syntactic.py

```diff
@@ -1298,6 +1320,11 @@ def _visit_With(self, stmt):
         if len(domain) == 1:
             domain = domain[0]
         body = self._visit(stmt.body)
+
+        if any(isinstance(l, (FunctionDef, ClassDef, Import)) for l in body):
+            errors.report("Unexpected object in a while block (function, class or import)",
+                    symbol=stmt, severity='error')
```

Comment:
There is also the case of an If Block
```python
if cond1:
    import file1
else:
    import fil2
```
This should not be allowed

Reply:
Now I have caught ifs. But this means that I am also catching `if __name__ == '__main__'`. Do you think we should disallow putting imports here or do I need to use a cleverer check?

Reply:
Maybe we should create another class called `MainIf` or something similar to catch that special case

Reply:
I'm not sure what to do if the user does something like
```python
if __name__ == '__main__':
    import file1
else:
    import file2

Reply:
In the semantic stage it is caught via
https://github.com/pyccel/pyccel/blob/b39fe9a411736a833d980674b7f6380278febf0f/pyccel/ast/core.py#L4465-L4473 

and turned into an instance of `Program`. But here we are checking the syntactic `expr` so we don't have that information yet

Reply:
I would prefer to disallow the condition:
```
if __name__ != '__main__':
```
I don't see a use case that would make it worth handling this properly

Reply:
We should probably ensure an error is raised in this case though

Reply:
For me it raises:
```
ERROR at annotation (semantic) stage
pyccel:
 |fatal [semantic]: tmp_sum.py [2,11]| Undefined variable (np)
```
Which seems like a very valid complaint as `f()` won't work as an import.

Reply:
https://peps.python.org/pep-0008/#imports

> Imports are always put at the top of the file, just after any module comments and docstrings, and before module globals and constants.

Reply:
I'm not sure how I feel about that :sweat_smile: . At least this should also work for imports I guess

Reply:
Yes, it is the same thing as
```python
def f():
    return x+1
x = 1
```
or 
```python
def f():
    return x+1
from file1 import x
``

Reply:
Maybe we should stick with the best practice

Reply:
So keep the error here and just move the import in the failing tests

Reply:
Fixed a8c984d02

Reply:
Not sure how I pushed this commit on its own. Here is the commit on this branch